servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/AbstractH2DuplexHandler.java

```diff
@@ -81,7 +81,7 @@ static void writeBuffer(ChannelHandlerContext ctx, Object msg, ChannelPromise pr
         if (buffer.readableBytes() > 0) {
             ctx.write(new DefaultHttp2DataFrame(encodeAndRetain(buffer), false), promise);
         } else {
-            promise.setSuccess();
+            ctx.write(EMPTY_BUFFER, promise);
```

Comment:
@idelpivnitskiy - this was done so we can preserve order of promise completion relative to writes. in some cases we may not have a reference to the "last write" and may do something like `write(empty buffer).whenComplete(close)`.

I saw in a commit message related to this change mention of a race between cancel and write on the write stream subscriber, but you made a recent fix which seems related https://github.com/apple/servicetalk/commit/4cb71f59afecc3c1cd1d6fc0f050cd98c0134eb9 ... let me know if you still feel we have a race with this approach as that maybe concerning in other scenarios.

Reply:
discussion moved to https://github.com/apple/servicetalk/pull/1850#discussion_r717859989